hkmc2/shared/src/main/scala/hkmc2/semantics/ImplicitResolver.scala

```
@@ -0,0 +1,336 @@
+  def resolveApp(base: Term, targs: Opt[List[Term]], argss: List[Term])(using ICtx) =
+  trace[Unit](s"Resolving App: $base; $targs; $argss", _ => s"~> $base; $targs; $argss"):
+    base match
+      case ModuleChecker.MethodDef(tdf) =>
+        given withTypeArgs: ICtx = tdf.tparams match
+          case S(tparams) => targs match
+            case S(targs) =>
+              if tparams.length != targs.length then
+                raise(ErrorReport(msg"Expected ${tparams.length.toString()} type arguments, " +
+                  msg"got ${targs.length.toString()}" -> base.toLoc :: Nil))
+              (tparams zip targs).foldLeft(ictx):
+                case (ictx, (tparam, targ)) => (tparam.sym, resolveType(targ)) match
+                  case (sym: TyParamSymbol, S(typ)) =>
+                    log(s"Resolving App with type arg ${sym} = $typ")
+                    ictx.withTypeArg(sym, typ)
+                  case _ => ictx
+            case N => tparams.foldLeft(ictx): 
+              case (ictx, tparam) => 
+                log(s"Resolving App with type arg ${tparam.sym} = ${Type.Unspecified}")
+                ictx.withTypeArg(tparam.sym, Type.Unspecified)
+          // No type parameters for the definition.
+          case N => ictx
+        val paramss = tdf.params
+        (paramss zip argss).foreach: 
+          case (ps, Term.Tup(args)) =>
+            // * TODO: Check the arity.
+            // if ps.paramCountUB
+            // then if ps.paramCountLB != args.length then
+            //   raise(ErrorReport(msg"Expected ${ps.paramCountLB.toString()} arguments, " +
+            //     msg"got ${args.length.toString()}" -> base.toLoc :: Nil))
+            // else if ps.paramCountLB > args.length then
+            //   raise(ErrorReport(msg"Expected at least ${ps.paramCountLB.toString()} arguments, " +
+            //     msg"got ${args.length.toString()}" -> base.toLoc :: Nil))
+            (ps.params zip args).foreach((p, arg) => resolveArg(p, arg)(base))
```

At least make the compiler crash instead of doing the wrong thing silently.
```suggestion
            assert(ps.params.sizeCompare(args) === 0)
            (ps.params zip args).foreach((p, arg) => resolveArg(p, arg)(base))
```

---

We cannot just assert if the argument lengths are equal because there are spread parameters.

I don't remember why I commented out the check, but seems it works fine. I just uncommented it and resolved any unit test problems in d510cd3e894f4233928212ff0f2c3fe07a4bc26b.

---

> I think we cannot just assert if the argument lengths are equal because there are spread parameters.

That was exactly the point: when you leave some checks commented and assume that something hold, you should back the assumption by a runtime check, otherwise you're going to do the wrong thing silently...